beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,504 @@
+		}
+
+		// earliestStart represents the time the slot starts, lowered by MAXIMUM_GOSSIP_CLOCK_DISPARITY.
+		// We lower the time by MAXIMUM_GOSSIP_CLOCK_DISPARITY in case system time is running slightly behind real time.
+		earliestStart := dv.clock.SlotStart(dataColumnSlot).Add(-maximumGossipClockDisparity)
+
+		// If the system time is still before earliestStart, we consider the column from a future slot and return an error.
+		if now.Before(earliestStart) {
+			return columnErrBuilder(ErrFromFutureSlot)
+		}
+	}
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SlotAboveFinalized() (err error) {
+	if ok, err := dv.results.cached(RequireSlotAboveFinalized); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSlotAboveFinalized, &err)
+
+	// Retrieve the finalized checkpoint.
+	finalizedCheckpoint := dv.fc.FinalizedCheckpoint()
+
+	// Compute the first slot of the finalized checkpoint epoch.
+	startSlot, err := slots.EpochStart(finalizedCheckpoint.Epoch)
+	if err != nil {
+		return columnErrBuilder(errors.Wrap(err, "epoch start"))
+	}
+
+	for _, dataColumn := range dv.dataColumns {
+		// Extract the data column slot.
+		dataColumnSlot := dataColumn.Slot()
+
+		// Check if the data column slot is after first slot of the epoch corresponding to the finalized checkpoint.
+		if dataColumnSlot <= startSlot {
+			return columnErrBuilder(ErrSlotNotAfterFinalized)
+		}
+	}
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) ValidProposerSignature(ctx context.Context) (err error) {
+	if ok, err := dv.results.cached(RequireValidProposerSignature); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireValidProposerSignature, &err)
+
+	for _, dataColumn := range dv.dataColumns {
+		// Extract the signature data from the data column.
+		signatureData := columnToSignatureData(dataColumn)
+
+		// Get logging fields.
+		fields := logging.DataColumnFields(dataColumn)
+		log := log.WithFields(fields)
+
+		// First check if there is a cached verification that can be reused.
+		seen, err := dv.sc.SignatureVerified(signatureData)
+		if err != nil {
+			log.WithError(err).Debug("Reusing failed proposer signature validation from cache")
+
+			columnVerificationProposerSignatureCache.WithLabelValues("hit-invalid").Inc()
+			return columnErrBuilder(ErrInvalidProposerSignature)
+		}
+
+		// If yes, we can skip the full verification.
+		if seen {
+			columnVerificationProposerSignatureCache.WithLabelValues("hit-valid").Inc()
+			continue
+		}
+
+		columnVerificationProposerSignatureCache.WithLabelValues("miss").Inc()
+
+		// Retrieve the parent state.
+		parentState, err := dv.parentState(ctx, dataColumn)
+		if err != nil {
+			return columnErrBuilder(errors.Wrap(err, "parent state"))
+		}
+
+		// Full verification, which will subsequently be cached for anything sharing the signature cache.
+		if err = dv.sc.VerifySignature(signatureData, parentState); err != nil {
+			return columnErrBuilder(errors.Wrap(err, "verify signature"))
+		}
+	}
+
+	return nil
+}
```

Comment:
I think the comment here is the same as the slot verification code re copy/paste, and probably for several other methods that are checking essentially the same things as we did for blobs. Maybe this could be something we come back to in a follow up PR...

Reply:
See https://github.com/OffchainLabs/prysm/pull/15232#discussion_r2095738879